sros2/sros2/errors.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+
+import lxml.etree
+
+
+class SROS2Error(Exception):
+    """Base class for all SROS2 exceptions."""
+
+
+class KeystoreError(SROS2Error):
+    """Base class for keystore-related exceptions."""
+
+
+class InvalidKeystoreError(KeystoreError):
+    """Exception raised when the keystore is invalid."""
+
+    def __init__(self, path: pathlib.Path):
+        self.path = path
+        super().__init__(f"'{str(path)}' is not a valid keystore")
+
+
+class KeystoreExistsError(KeystoreError):
+    """Exception raised when trying to create a keystore that already exists."""
+
+    def __init__(self, path: pathlib.Path):
+        self.path = path
+        super().__init__(f"keystore '{str(path)}' already exists")
+
+
+class EnclaveError(SROS2Error):
+    """Base class for enclave-related exceptions."""
+
+
+class InvalidEnclaveNameError(EnclaveError):
+    """Exception raised when an enclave name is invalid."""
+
+    def __init__(self, name: str):
+        self.name = name
+        super().__init__(f"'{name}' is not a valid enclave name")
+
+
+class XmlError(SROS2Error):
```

Comment:
This base may be a little confusing wrt actual `lxml` error. I would rename it or even remove it and have child classes inherit directly from `SROS2Error`.

Reply:
lxml has parser errors and schema errors, this is collecting all the XML-related errors for SROS2. I don't really see how that's confusing, and I believe there's value in that organization. Can you expand on what you think might be confusing?

Reply:
Oh it is simply that reading `XmlError` I would expect the class to come from the `lxml` library. I like the organization but not the naming so much, I know that's nitpick. How about,
```suggestion
class SROS2XmlError(SROS2Error):
```

Reply:
Eh... I'm not a big fan. Think of how it will be used:

```python3
import sros2.errors
# ...
sros2.errors.SROS2XmlError
```

or

```python3
from sros2.errors import SROS2XmlError
```

There's a lot of "sros2" going on there; I don't think adding it here has meaning, just typing. If there was actually an `XmlError` where this might clash I'd feel differently, but there's not.